src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -196,6 +197,27 @@ public virtual void ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags an
         {
             throw new NotImplementedException();
         }
+
+        /// <summary>
+        /// Attempts to compute or get the cached value provided by the given <paramref name="valueProvider"/> for the given <paramref name="key"/>.
+        /// Re-using the same <paramref name="valueProvider"/> instance across analyzer actions and/or analyzer instances can improve the overall analyzer performance by avoiding recomputation of the values.
+        /// </summary>
+        /// <typeparam name="TValue">The type of the value associated with the key</typeparam>
+        /// <param name="key"><see cref="SourceText"/> for which the value is queried</param>
+        /// <param name="valueProvider">Provider that computes the underlying value associated with the key</param>
+        /// <param name="value">Value associated with the key</param>
+        /// <returns>Returns true on success, false otherwise</returns>
+        public bool TryGetValue<TValue>(SourceText key, SourceTextValueProvider<TValue> valueProvider, out TValue value)
```

Comment:
This is the core API/functionality being added - analyzers will be able to compute and cache per-SourceText or per-SyntaxTree values and use it across analyzer actions and also across analyzer instances. They don't need to be concerned about cache lifetime or management.